File: typescript/src/capture.ts
import { buildEvent, currentSpan, newSpan, resolveStorage } from "./context.ts";
import { toJsonable } from "./serialization.ts";
import type { Storage } from "./storage/base.ts";
import type { TokenUsage } from "./types.ts";
import { createId, durationMs, utcNowIso } from "./utils.ts";

export async function captureLlmCall(input: {
  name: string;
  provider: string;
  model?: string;
  prompt: unknown;
  response?: unknown;
  startedAt?: string;
  endedAt?: string;
  error?: unknown;
  usage?: TokenUsage;
  tags?: string[];
  metadata?: Record<string, unknown>;
  storage?: Storage | string;
}): Promise<void> {
  const active = currentSpan();
  const span = active ?? newSpan(input.name);
  const event = buildEvent({
    kind: "llm",
    name: input.name,
    span,
    provider: input.provider,
    model: input.model,
    prompt: input.prompt,
    response: input.response,
    tags: input.tags,
    metadata: input.metadata,
  });

  event.spanId = createId();
  event.parentSpanId = active?.spanId ?? span.parentSpanId;
  event.createdAt = input.startedAt ?? utcNowIso();
  event.endedAt = input.endedAt ?? utcNowIso();
  event.latencyMs = durationMs(event.createdAt, event.endedAt);
  event.response = toJsonable(input.response);
  event.error = input.error
    ? input.error instanceof Error
      ? `${input.error.name}: ${input.error.message}`
      : String(input.error)
    : undefined;
  event.inputTokens = input.usage?.inputTokens;
  event.outputTokens = input.usage?.outputTokens;

  await resolveStorage(input.storage).writeEvent(event);
}

File: src/blackboxd/capture.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any
from uuid import uuid4

from blackboxd.context import build_event, current_span, new_span, resolve_storage
from blackboxd.models import TokenUsage
from blackboxd.serialization import to_jsonable
from blackboxd.storage.base import Storage


def utc_now() -> datetime:
    return datetime.now(timezone.utc)


def capture_llm_call(
    *,
    name: str,
    provider: str,
    model: str | None,
    prompt: Any,
    started_at: datetime | None = None,
    ended_at: datetime | None = None,
    response: Any = None,
    error: Exception | None = None,
    usage: TokenUsage | None = None,
    tags: list[str] | None = None,
    metadata: dict[str, Any] | None = None,
    storage: Storage | str | None = None,
) -> None:
    active = current_span()
    synthetic_span = active or new_span(name)
    llm_span_id = str(uuid4())
    event = build_event(
        kind="llm",
        name=name,
        span=synthetic_span,
        provider=provider,
        model=model,
        prompt=prompt,
        response=response,
        tags=tags,
        metadata=metadata,
    )
    event.span_id = llm_span_id
    event.parent_span_id = active.span_id if active else synthetic_span.parent_span_id
    event.created_at = started_at or utc_now()
    event.finish(
        ended_at=ended_at or utc_now(),
        response=to_jsonable(response),
        error=f"{type(error).__name__}: {error}" if error else None,
    )
    event.apply_usage(usage)
    resolve_storage(storage).write_event(event)
